android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/BroadcastActivity.java

```diff
@@ -0,0 +1,207 @@
+/*
+MainActivity.java
+
+This is the main entry point for the application
+It is the primary ui interface
+It also creates the singletons used for the main app logic
+like sharing a URL via bluetooth, and interfacing with Drive
+It also contains the logic for ensuring permissions
+and other account requirements are satisfied
+*/
+
+package org.physical_web.physicalweb;
+
+import android.annotation.TargetApi;
+import android.app.Activity;
+import android.bluetooth.BluetoothAdapter;
+import android.bluetooth.BluetoothManager;
+import android.content.Context;
+import android.content.Intent;
+import android.os.Build;
+import android.util.Log;
+import android.util.Patterns;
+import android.widget.Toast;
+
+import org.physical_web.physicalweb.ble.UriBeacon;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.regex.Matcher;
+
+
+/**
+ * Handles share intent.
+ */
+public class BroadcastActivity extends Activity {
+
+    private static final String TAG  = "BroadcastActivity";
+    private static final int ENABLE_BLUETOOTH_REQUEST_ID = 1;
+    public static final int MAX_URI_LENGTH = 18;
+    private String displayUrl;
+
+
+    // callbacks
+    /////////////////////////////////
+    @Override
+    protected void onActivityResult (int requestCode, int resultCode, Intent data) {
+      Log.d(TAG, "onActivityResult");
+      if (requestCode == ENABLE_BLUETOOTH_REQUEST_ID && resultCode == -1) {
+        if (checkLE()){
+            preBroadcast();
+        }
+            finish();
+      } else {
+        Toast.makeText(this, "Turn Bluetooth on to broadcast", Toast.LENGTH_LONG).show();
+        finish();
+      }
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        if (Build.VERSION.SDK_INT < 21){
+            Toast.makeText(this, "Bluetooth LE advertising is not supported (API)",
+                Toast.LENGTH_LONG).show();
+            return;
+        }
+        if (!checkIfBluetoothIsEnabled()){
+            requestBluetooth();
+        } else {
+            if (checkLE()){
```

Comment:
I'm thinking this should be its own method since you have the same code up above for when the permission loop is complete.